Generator/Sources/NeedleFramework/Generating/Serializers/OutputSerializer.swift

```diff
@@ -37,23 +37,74 @@ class OutputSerializer: Serializer {
+        // codebases. This tells us how many levels deep the tree goes.
+        var maxLevel: Int = 1
 
         let providersSection = providers
             .map { (provider: SerializedProvider) in
-                provider.content
+                if let providerMaxLevel = Int(provider.attributes["maxLevel"] ?? "0") {
+                    if providerMaxLevel > maxLevel {
+                        maxLevel = providerMaxLevel
+                    }
+                }
+                if let factoryName = provider.attributes["factoryName"] {
+                    if registeredFactories.contains(factoryName) {
+                        return ""
+                    }
+                    registeredFactories.insert(factoryName)
+                }
+                return provider.content
             }
             .joined()
+
+        let traversalHelpers = (1...maxLevel).map { num in
+            return """
+            private func parent\(num)(_ component: NeedleFoundation.Scope) -> NeedleFoundation.Scope {
+                return component\(String(repeating: ".parent", count: num))
+            }
+            """
+        }.joined(separator: "\n\n")
         
         let needleDependenciesHash = needleVersionHash.map { return "\"\($0)\""} ?? "nil"
 
         let importsJoined = imports.joined(separator: "\n")
 
+        // With Swift 5.6 and an amd64 target, having a function body that is
+        // thousands of lines long causes a severe compile performance
+        // regression. We were seeing compilation take 4.5x what an x86_64
+        // target would take. As a result, this code splits the calls to
+        // register the dependencies into functions around 50 lines long.
+        // Through some basic testing, this seemed to produce the best results.
+        let linesPerHelper = 50
+        var registrationHelperFuncs: Array<String> = []
+        let registrations: Array<String> = providers
+            .map { (provider: SerializedProvider) in
+                provider.registration
+            }
+            .filter {
+                !$0.isEmpty
+            }
+        let registrationBody = stride(from: 0, to: registrations.count, by: linesPerHelper)
+            .map {
+                let helperBody = registrations[$0 ..< Swift.min($0 + linesPerHelper, registrations.count)]
+                    .joined()
+                    .replacingOccurrences(of: "\n", with: "\n    ")
+                    .trimmingCharacters(in: .whitespacesAndNewlines)
+                let regNum = ($0 / linesPerHelper) + 1
+                registrationHelperFuncs.append("""
+                private func register\(regNum)() {
+                    \(helperBody)
+                }
+                """)
+                return "register\(regNum)()"
+            }
+            .joined(separator: "\n    ")
+        let registrationHelpers = registrationHelperFuncs.joined(separator: "\n\n")
```

Comment:
nit: can we put this block in `DependencyProviderRegistrationSerializer` serialize function?

Reply:
I don't believe so. `DependencyProviderRegistrationSerializer` generates an individual registration. Once all of the registrations have been accumulated, then we need to chunk them up so we don't have functions with more than around 50 statements.